### PR TITLE
Hughsie/meteorlake prep

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@superm1 mind if I cherry pick `Support partial bytes in Rust structs` into main? I've got another user.

**Comment by superm1** [comment]
> @superm1 mind if I cherry pick `Support partial bytes in Rust structs` into main? I've got another user.

Go ahead.

**Review by superm1 — APPROVED** [review]

**Comment by hughsie** [comment]
@superm1 the CSME 18.x and 19.x stuff is still WIP -- but the AMT is a nice cleanup.

**Comment by superm1** [comment]
> @superm1 the CSME 18.x and 19.x stuff is still WIP -- but the AMT is a nice cleanup.

Cool will make landing more I'm future easier and cleaner then too.

**Comment by hughsie** [comment]
Intel doesn't want to help with this feature, so I guess I'll merge it based on my best guesses. @mrhpearson if you want to test this branch that'd be great, otherwise I can do you a fedora copr test build if that would be helpful.

**Comment by mrhpearson** [comment]
> Intel doesn't want to help with this feature, so I guess I'll merge it based on my best guesses. @mrhpearson if you want to test this branch that'd be great, otherwise I can do you a fedora copr test build if that would be helpful.

By happenstance I did bug Intel for support on this earlier this morning. Not expecting much though to be honest...email thread has been a bit torturous.

Built and tested your branch on X1 Carbon G12 - let me know if I'm doing anything wrong from the below. Are we just skipping the BootGuard status on these devices?

```
(fwupd) banther@x1c12:~/projects/fwupd$ ./venv/bin/fwupdtool security
Loading…                 [****                                   ]17:41:47.029 FuPluginMm           failed to add instance ID USB\VID_2C7C&PID_0310&REV_0318&NAME_EM05CNF
Loading…                 [*******************                    ]17:41:47.698 FuPluginIntelMe      failed to get public key using file-id 0x40002300: generic failure [0xb]
Loading…                 [************************************** ]
WARNING: This package has not been validated, it may not work properly.
Host Security ID: HSI:3! (v2.0.0)

HSI-1
✔ BIOS firmware updates:         Enabled
✔ MEI key manifest:              Valid
✔ Platform debugging:            Disabled
✔ SPI write:                     Disabled
✔ SPI lock:                      Enabled
✔ SPI BIOS region:               Locked
✔ Supported CPU:                 Valid
✔ TPM empty PCRs:                Valid
✔ TPM v2.0:                      Found
✔ UEFI bootservice variables:    Locked
✔ UEFI platform key:             Valid

HSI-2
✔ BIOS rollback protection:      Enabled
✔ IOMMU:                         Enabled
✔ Platform debugging:            Locked
✔ TPM PCR0 reconstruction:       Valid

HSI-3
✔ CET Platform:                  Supported
✔ Pre-boot DMA protection:       Enabled
✔ Suspend-to-idle:               Enabled
✔ Suspend-to-ram:                Disabled

HSI-4
✔ SMAP:                          Enabled
✘ Encrypted RAM:                 Not supported

Runtime Suffix -!
✔ fwupd plugins:                 Untainted
✔ Linux kernel:                  Untainted
✘ CET OS Support:                Not supported
✘ Linux kernel lockdown:         Disabled
✘ Linux swap:                    Unencrypted
✘ UEFI secure boot:              Disabled
```

**Comment by hughsie** [comment]
Hey Mark. I've just got an email from Pragy myself about this very topic :) She'll ask you to escalate, I can elaborate via email if required.

I'm a bit confused why you don't get any CSME entries at all -- I was expecting `fu_mei_detect_family()` to return `CSME18`. If you can debug that a bit that would be great.

**Comment by mrhpearson** [comment]
Found part of the problem - left a comment in the commit, but in fu-pci-mei-plugin.c, function add_security_attrs the check to see if hfsts6 == NULL should be hfsts5 (line 712)
With that fixed I'm now seeing the Intel BootGuard options in the description.
On my system they're showing as Disabled/Invalid - which I _think_ is wrong....but as my system is a proto may be right. I need to go and do some register maths, and checking to confirm. Will dig a bit deeper, hopefully this afternoon.
Mark

**Comment by hughsie** [comment]
> function add_security_attrs the check to see if hfsts6 == NULL should be hfsts5

Good catch! You're basically an official fwupd developer now, right? :)

> On my system they're showing as Disabled/Invalid - which I think is wrong....but as my system is a proto

Can you share the `-vv` output? That should give you the register bits on the console.

**Comment by mrhpearson** [comment]
OK - with the other change (comment in commit - line 376 - typo on checking return value) I think it's working.

Unfortunately on my system the Bootguard OTP Fuse settings are not set - and fwupd is correctly returning that they are disabled. It would be nice to test on a shipping system that it works - but the code all looks good so I think it's safe.

**Comment by mrhpearson** [comment]
> Good catch! You're basically an official fwupd developer now, right? :)

Yeah! I add debug in the code and everything. It's been the best Monday morning I've had a long time ;)

**Comment by hughsie** [comment]
I don't see the comment, but I think you mean something like this, right?

            /* disabled at runtime? */
    -       if (fu_mei_csme18_hfsts5_get_valid(hfsts5)) {
    +       if (!fu_mei_csme18_hfsts5_get_valid(hfsts5)) {

**Comment by mrhpearson** [comment]
> I don't see the comment, but I think you mean something like this, right?
> 
> ```
>         /* disabled at runtime? */
> -       if (fu_mei_csme18_hfsts5_get_valid(hfsts5)) {
> +       if (!fu_mei_csme18_hfsts5_get_valid(hfsts5)) {
> ```

Yes - exactly that. Well figured out if my comment isn't showing up.

I'll re-run the test when you push the fix

Mark

**Review by mrhpearson — COMMENTED** [review]
Ah - my bad...I never actually submitted for review.
Doing this just for completeness - I know you've already figured it out

File: plugins/pci-mei/fu-pci-mei-plugin.c [hunk 1]
```diff
@@ -308,14 +342,38 @@ fu_plugin_add_security_attrs_bootguard_enabled(FuPlugin *plugin, FuSecurityAttrs
...
+
+	/* no device */
+	if (self->pci_device == NULL) {
+		fwupd_security_attr_set_result(attr, FWUPD_SECURITY_ATTR_RESULT_NOT_FOUND);
 		return;
 	}
 
 	/* disabled at runtime? */
-	if (self->hfsts6.fields.boot_guard_disable) {
+	if (fu_mei_csme18_hfsts5_get_valid(hfsts5)) {
```
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/pci-mei/fu-pci-mei-plugin.c]: Fixed, thanks.

**Pull request merged**